orchestrator/careplancontributor/ehr/notifier.go

```diff
@@ -0,0 +1,373 @@
+//
+// Parameters:
+//   - tasks: A list of FHIR tasks.
+//
+// Returns:
+//   - []string: A list of questionnaire input references.
+func findQuestionnaireInputs(tasks []fhir.Task) []string {
+	var questionnaireRefs []string
+	for _, task := range tasks {
+		questionnaireRefs = append(questionnaireRefs, fetchTaskInputs(task)...)
+	}
+	return questionnaireRefs
+}
+
+// findQuestionnaireOutputs extracts questionnaire output references from a list of tasks.
+//
+// Parameters:
+//   - tasks: A list of FHIR tasks.
+//
+// Returns:
+//   - []string: A list of questionnaire output references.
+func findQuestionnaireOutputs(tasks []fhir.Task) []string {
+	var questionnaireResponseRefs []string
+	for _, task := range tasks {
+		questionnaireResponseRefs = append(questionnaireResponseRefs, fetchTaskOutputs(task)...)
+	}
+	return questionnaireResponseRefs
+}
+
+// fetchTaskOutputs extracts questionnaire response references from a task's outputs.
+//
+// Parameters:
+//   - task: A FHIR task.
+//
+// Returns:
+//   - []string: A list of questionnaire response references.
+func fetchTaskOutputs(task fhir.Task) []string {
+	var questionnaireResponseRefs []string
+	if task.Output != nil {
+		for _, output := range task.Output {
+			if output.ValueReference != nil &&
+				output.ValueReference.Reference != nil {
+				matchesType := isOfOutputType(output, "QuestionnaireResponse")
+				if matchesType {
+					reference := *output.ValueReference.Reference
+					if !slices.Contains(questionnaireResponseRefs, reference) {
+						questionnaireResponseRefs = append(questionnaireResponseRefs, reference)
+					}
+				}
+			}
+		}
+	}
+	return questionnaireResponseRefs
+}
+
+// isOfOutputType checks if a task output is of a specific type.
+//
+// Parameters:
+//   - output: The task output to check.
+//   - typeName: The type name to check against.
+//
+// Returns:
+//   - bool: True if the output is of the specified type, false otherwise.
+func isOfOutputType(output fhir.TaskOutput, typeName string) bool {
+	matchesType := false
+	if output.ValueReference.Type != nil {
+		matchesType = *output.ValueReference.Type == typeName
+	} else if output.ValueReference.Reference != nil {
+		matchesType = strings.HasPrefix(*output.ValueReference.Reference, fmt.Sprintf("%s/", typeName))
+	}
+	return matchesType
+}
+
+// isOfInputType checks if a task input is of a specific type.
+//
+// Parameters:
+//   - input: The task input to check.
+//   - typeName: The type name to check against.
+//
+// Returns:
+//   - bool: True if the input is of the specified type, false otherwise.
+func isOfInputType(input fhir.TaskInput, typeName string) bool {
+	matchesType := false
+	if input.ValueReference.Type != nil {
+		matchesType = *input.ValueReference.Type == typeName
+	} else if input.ValueReference.Reference != nil {
+		matchesType = strings.HasPrefix(*input.ValueReference.Reference, fmt.Sprintf("%s/", typeName))
+	}
+	return matchesType
```

Comment:
If you make 1 function for `valueReference` type instead, you don't need to duplicate this

Reply:
Refactored